spec/predicates/vsa.md

```diff
@@ -1,11 +1,358 @@
+
+<a id="verifier.version"></a>
+`verifier.version` _map (string->string), optional_
+
+> Map of names of components of the verification platform to their version.
+
+<a id="timeVerified"></a>
+`timeVerified` _string ([Timestamp]), optional_
+
+> Timestamp indicating what time the verification occurred.
+
+<a id="resourceUri"></a>
+`resourceUri` _string ([ResourceURI]), required_
+
+> URI that identifies the resource associated with the artifact being verified.
+
+<a id="policy"></a>
+`policy` _object ([ResourceDescriptor]), required_
+
+> Describes the policy that the `subject` was verified against.
+>
+> The entry MUST contain a `uri`.
+
+<a id="inputAttestations"></a>
+`inputAttestations` _array ([ResourceDescriptor]), optional_
+
+> The collection of attestations that were used to perform verification.
+> Conceptually similar to the `resolvedDependencies` field in [SLSA Provenance].
+>
+> This field MAY be absent if the verifier does not support this feature.
+> If non-empty, this field MUST contain information on _all_ the attestations
+> used to perform verification.
+>
+> Each entry MUST contain a `digest` of the attestation and SHOULD contains a
+> `uri` that can be used to fetch the attestation.
+
+<a id="verificationResult"></a>
+`verificationResult` _string, required_
+
+> Either “PASSED” or “FAILED” to indicate if the artifact passed or failed the policy verification.
+
+<a id="verifiedProperties"></a>
+`verifiedProperties` _array (_string), required_
+
+> Indicates the properties verified for the artifact (and not
+> its dependencies).
+
+<a id="dependencyProperties"></a>
+`dependencyProperties` _object, optional_
+
+> A count of the dependencies verified to have each property.
+>
+> Map from _string_ to the number of the artifact's _transitive_ dependencies
+> that were verified at the indicated level. Absence of a given value
+>  MUST be interpreted as reporting _0_ dependencies with that value.
+
+## Example
+
+WARNING: This is just for demonstration purposes.
+
+```jsonc
+"_type": "https://in-toto.io/Statement/v1",
+"subject": [{
+  "name": "out/example-1.2.3.tar.gz",
+  "digest": {"sha256": "5678..."}
+}],
+
+// Predicate
+"predicateType": "https://in-toto.io/attestation/verification_summary/v1",
+"predicate": {
+  "verifier": {
+    "id": "https://example.com/publication_verifier",
+    "version": {
+      "slsa-verifier-linux-amd64": "v2.3.0",
+      "slsa-framework/slsa-verifier/actions/installer": "v2.3.0"
+    }
+  },
+  "timeVerified": "1985-04-12T23:20:50.52Z",
+  "resourceUri": "https://example.com/example-1.2.3.tar.gz",
+  "policy": {
+    "uri": "https://example.com/example_tarball.policy",
+    "digest": {"sha256": "1234..."}
+  },
+  "inputAttestations": [
+    {
+      "uri": "https://example.com/provenances/example-1.2.3.tar.gz.intoto.jsonl",
+      "digest": {"sha256": "abcd..."}
+    }
+  ],
+  "verificationResult": "PASSED",
+  "verifiedProperties": ["SLSA_BUILD_LEVEL_3"],
```

Comment:
The relationship between the policy and verification results that I find myself expecting is:

* `policy` is a list of policies that are considered in the VSA
* `verificationResult` is the net result -- whether all policies pass or fail
* `verificationProperties` is some indication about the result from each policy taken into account (i.e. whether it is pass or fail)

What might be better, however, is to have keys following the pattern:
*  `policies` - the list of all policies that have been applied to create this VSA
*  `result` - the ultimate result of all policies combined
*  `policyResults` - the individual result of each policy applied. There would have to be easy mapping from each policy to its result, but this could potentially be done by assigning a unique name for each policy so that the full uri/digest don't have to be repeated.

Outside of the SLSA specification, it might be harder to realize value from a field like `verifiedProperties`.

Reply:
> the ultimate result of all policies combined

It's not clear to me how this should actually be calculated.

For my intended consumption of this, trust in a VSA will be based on a 1:1 mapping of trust in a policy. If I revoke trust in a policy (i.e. just to update the policy). I will revoke trust in all VSA's generated from that policy. Not arguing against the use case, I don't think we would ever use multiple policies in a single VSA. I'm curious to hear from others.

Reply:
Could we consider `policy`+`verificationResult` vs. `verifiedProperties` as two sibling structures to convey trust? i.e.:

- One could decide to delegate trust on a specific policy (if they know its content meets their needs), and just verify that: 1) the `policy` that they are interested has been verified; and 2) the overall `verificationResult` is a PASS.

- Alternatively, one could decide they accept any artifact that meets a certain set of `verifiedProperties`, e.g. SLSA_BUILD_3 and SLSA_SOURCE_2 (imaginary), and not caring what exact policy triggered the verification of those properties. In that case, they will only be looking up their expected keys in `verifiedProperties`.

Of course, the latter requires the keys in `verifiedProperties` to be arbitrated (i.e. having a higher-level protocol like SLSA). Otherwise any verifier can put any random text in it that can be misunderstood by other verifiers.

---
OTOH, if we remove the `verificationResult` as @MarkLodato suggests [here](#discussion_r1394481717), then `verifiedProperties` becomes the only facility to convey trust.

Reply:
> For my intended consumption of this, trust in a VSA will be based on a 1:1 mapping of trust in a policy. If I revoke trust in a policy (i.e. just to update the policy). I will revoke trust in all VSA's generated from that policy. Not arguing against the use case, I don't think we would ever use multiple policies in a single VSA. I'm curious to hear from others.

@jkjell if policy A delegates to policy B, and therefore verification of policy A results in VSA A and VSA B, I don't think we can claim that revoking policy A automatically revokes policy B, at least in the in-toto sublayout context. Policy B may be delegated to by C, D, E, etc as well, so IMO VSA B is fine to continue to trust in isolation. Instead, as your verification starts at policy A (the root policy), you'd not consider VSA B in that instance of the workflow, rather than _revoking_ VSA B.